addons/sale/models/res_partner.py

```diff
@@ -1,29 +1,21 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from openerp.osv import fields,osv
-from openerp.addons.base.res.res_partner import WARNING_MESSAGE, WARNING_HELP
+from odoo import fields, models
+from odoo.addons.base.res.res_partner import WARNING_MESSAGE, WARNING_HELP
 
-class res_partner(osv.osv):
-    _inherit = 'res.partner'
 
-    def _sale_order_count(self, cr, uid, ids, field_name, arg, context=None):
-        res = dict(map(lambda x: (x,0), ids))
-        # The current user may not have access rights for sale orders
-        try:
-            for partner in self.browse(cr, uid, ids, context):
-                res[partner.id] = len(partner.sale_order_ids) + len(partner.mapped('child_ids.sale_order_ids'))
-        except:
-            pass
-        return res
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
 
-    _columns = {
-        'sale_order_count': fields.function(_sale_order_count, string='# of Sales Order', type='integer'),
-        'sale_order_ids': fields.one2many('sale.order', 'partner_id', 'Sales Order'),
-        'sale_warn': fields.selection(WARNING_MESSAGE, 'Sales Order', default='no-message', help=WARNING_HELP, required=True),
-        'sale_warn_msg': fields.text('Message for Sales Order'),
-    }
+    sale_order_count = fields.Integer(compute='_compute_sale_order_count', string='# of Sales Order')
+    sale_order_ids = fields.One2many('sale.order', 'partner_id', 'Sales Order')
+    sale_warn = fields.Selection(WARNING_MESSAGE, 'Sales Order', default='no-message', help=WARNING_HELP, required=True)
+    sale_warn_msg = fields.Text('Message for Sales Order')
 
-    _defaults = {
-        'sale_warn': 'no-message',
-    }
+    def _compute_sale_order_count(self):
+        SaleOrder = self.env['sale.order']
+        for partner in self:
+            partner.sale_order_count = SaleOrder.search_count([
```

Comment:
I have a doubt: does child_of take into account the current partner, or only its childs ?